service/matching/taskListManager.go

```diff
@@ -410,12 +414,23 @@ func (c *taskListManagerImpl) getTask(ctx context.Context) (*getTaskResult, erro
 	defer timer.Stop()
 
 	pollerID, ok := ctx.Value(pollerIDKey).(string)
+
 	childCtx := ctx
+	// We need to set a shorter timeout than the original ctx; otherwise, by the time ctx deadline is
+	// reached, instead of emptyTask, context timeout error is returned to the frontend by the rpc stack,
+	// which counts against our SLO. By shortening the timeout by a very small amount, the emptyTask can be
+	// returned to the handler before a context timeout error is generated.
+	if deadline, ok := childCtx.Deadline(); ok {
+		var cancel context.CancelFunc
+		childCtx, cancel = context.WithTimeout(ctx, deadline.Sub(time.Now())-returnEmptyTaskTimeBudget)
+		defer cancel()
```

Comment:
How/When this context gets canceled seems weird to me. There is a defer cancel here and there is also one below. I think we want to make this more simple. How about the child context should always have a deadline and a cancel function. If the parent has a deadline then do what you are doing now. If the parent does not have a deadline then you can use `c.config.LongPollExpirationInterval()`

That way you can just have the function do `defer cancel()` in one place.

Reply:
In fact think through if we even need the timer with this approach at all?